cpp/src/io/parquet/page_string_decode.cu

```diff
@@ -0,0 +1,1157 @@
+  while (!s->error && (s->input_value_count < s->num_input_values || s->src_pos < s->nz_count)) {
+    int src_pos = s->src_pos;
+    // target pos for level decoding
+    int target_pos = min(src_pos + decode_block_size, s->nz_count + decode_block_size);
+
+    if (t < 32) {
+      // decode repetition and definition levels.
+      // - update validity vectors
+      // - updates offsets (for nested columns)
+      // - produces non-NULL value indices in s->nz_idx for subsequent decoding
+      gpuDecodeLevels<lvl_buf_size>(s, sb, target_pos, rep, def, t);
+    } else if (t < out_thread0) {
+      // skipped_leaf_values will always be 0 for flat hierarchies.
+      uint32_t src_target_pos = target_pos + skipped_leaf_values;
+
+      // WARP1: Decode dictionary indices, booleans or string positions
+      if (s->dict_base) {
+        src_target_pos = gpuDecodeDictionaryIndices<false>(s, sb, src_target_pos, t & 0x1f).first;
+      } else {
+        gpuInitStringDescriptors<false>(s, sb, src_target_pos, t & 0x1f);
+      }
+      if (t == 32) { *(volatile int32_t*)&s->dict_pos = src_target_pos; }
+    }
+    __syncthreads();
+
+    // target_pos for value decoding
+    target_pos = min(s->nz_count, target_pos);
+
+    // if this is the first page, then the first non-null entry will have an offset of 0.
+    // pages that start with a run of nulls will have repeated 0 values, so for the fixing
+    // of null offsets done at the end, we need to know the last index that should be 0.
+    if (t == 0 && s->page.str_offset == 0 && first_non_null == -1) {
+      for (int i = src_pos; i < target_pos; i++) {
+        int dst_pos = sb->nz_idx[rolling_index(i)];
+        if (!has_repetition) { dst_pos -= s->first_row; }
+        if (dst_pos >= 0) {
+          first_non_null = dst_pos;
+          break;
+        }
+      }
+    }
+
+    // Decode values
+    src_pos += t;
+
+    // the position in the output column/buffer
+    int dst_pos = sb->nz_idx[rolling_index(src_pos)];
+
+    // for the flat hierarchy case we will be reading from the beginning of the value stream,
+    // regardless of the value of first_row. so adjust our destination offset accordingly.
+    // example:
+    // - user has passed skip_rows = 2, so our first_row to output is 2
+    // - the row values we get from nz_idx will be
+    //   0, 1, 2, 3, 4 ....
+    // - by shifting these values by first_row, the sequence becomes
+    //   -1, -2, 0, 1, 2 ...
+    // - so we will end up ignoring the first two input rows, and input rows 2..n will
+    //   get written to the output starting at position 0.
+    //
+    if (!has_repetition) { dst_pos -= s->first_row; }
+
+    // need to do this before we branch on src_pos/dst_pos so we don't deadlock
+    // choose a character parallel string copy when the average string is longer than a warp
+    using cudf::detail::warp_size;
+    auto const use_char_ll =
+      s->page.num_valids > 0 && (s->page.str_bytes / s->page.num_valids) >= warp_size;
```

Comment:
It can. But divide by zero is ok on device?  I added a print statement when `num_valids` is zero, but removing the check for `num_valids > 0` didn't crash.

Reply:
Disregard, I completely missed the `s->page.num_valids > 0` part of the expression, you got it covered :D
I'm actually not sure what would happen if an int is divided by zero on the device.